### PR TITLE
Port to Python 3

### PR DESCRIPTION
**Pull request by vkareh**
This adds conditional macros to allow compiling for either Python 2 or Python 3. Tested today on Ubuntu MATE daily build and works well for me on a ported version of mate-menu. Might need extra testing to ensure compatibility with other menu systems.

### CODE REVIEW

**Comment by raveit65** [comment]
Before i test something, it is more important that it works with our applications.
Are normal menus or mozo working with this PR?

**Comment by flexiondotorg** [comment]
@vkareh Thanks for working on this. I'll test it when I have access to a computer again in a week.

**Comment by monsta** [comment]
I don't really want to go this way. I'd prefer to add GObject introspection support instead and then drop these old bindings.

**Comment by vkareh** [comment]
@raveit65 - good call. This works with mate-menu (when used with this PR: https://github.com/ubuntu-mate/mate-menu/pull/28). I will test with mozo (I'm assuming it'll need to be ported to python3 as well)

@monsta - I don't know much about GObject Introspection, but I'm reading up on it and seems like it would probably be the best way moving forward. I'd have to research what it would take to do this.

**Comment by raveit65** [comment]
I don't like the tons of introspection build warnings in our code, and nobody takes care of it since years......
@vkareh 
Thanks for the hint. Personal i use normal standard menus, so for better understanding.
This python code is used mintmenu/mate-menu and mozo, anything left?

**Comment by vkareh** [comment]
@raveit65 - hmm... those are the only ones I'm aware of. A quick search on GitHub shows a handful of small projects, most of them forks of mintmenu anyway. Just to note: this PR still allows to build with Python2. I'm wondering if I should probably change it so that the user can pass a flag during build.

I'll also need to port mintmenu and mozo to python3

**Comment by raveit65** [comment]
> We have a request from Monsta to do this a different way, I'll wait on that

Honestly, i want to hear more pros ans cons before we decide something.

> Just to note: this PR still allows to build with Python2. I'm wondering if I should probably change it so that the user can pass a flag during build.

I am fine with that, as it doesn't made any different if the code runs with python3 or 2 if it works.
And if someone like to port it  gir , this can be done late.
At least i like to hear @XRevan86 opinion. He is the package mantainer and knows more about python and gir than most of us here.

**Comment by raveit65** [comment]
At least fedora wants to finish python3 migration, sooner or later.
https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/message/TQZXTEYAPUPAODWKG2K2NGVZJNAL5ZDR/
So, using python3 will make my live as fedora maintainer easier. ;-)

**Comment by flexiondotorg** [comment]
Ubuntu MATE will fund this work. So let's agree the best way to proceed ☺️

**Comment by XRevan86** [comment]
```AM_PATH_PYTHON(3.6)``` – but why? It looks like compatible with both Pythons except for this part which demands Python 3.6+.
Also I'd say a better practice would be to make the newer API a default:
```c
#if PY_MAJOR_VERSION < 2
+	#define PyUnicode_FromString PyString_FromString
+	#define PyUnicode_Check PyString_Check
+	#define PyUnicode_AsUTF8 PyString_AsString
+#endif
```

**Comment by vkareh** [comment]
@XRevan86 - I followed your recommendation about making Python 3 the default. I've pushed new changes to the pull request. Also, I've been thinking about this and you probably have a better formed opinion about this: I changed the build process to check for `python3` first even if the system's default is a lower version. It still allows 2.7+, but it would default to the higher of the two versions. I'm second-guessing myself now and thinking this could break things for apps that use this library, but expect to be compiled with whatever default the distro has... thoughts? Should I remove the `_AM_PYTHON_INTERPRETER_LIST` override and let the distro decide?

**Comment by flexiondotorg** [comment]
@monsta While this might not be the preferred end goal, could we not use this an as intermediate step? For example, Ubuntu MATE has few applications that use python2 it would be great to get everything transitioned to python3 so the iso size can be better managed. And as python2 migrates to community maintained status only python3 will actively see security updates in Ubuntu.

**Comment by XRevan86** [comment]
@vkareh, um, that's not what I meant… I meant that in the code the code that stands first should be for python3.

**Comment by vkareh** [comment]
@XRevan86 - I did that too: https://github.com/mate-desktop/mate-menus/pull/54/commits/37565ade333b1360add497889825db764a99af13#diff-6a17f61bec63b7ad05bd90563055aa0f

**Comment by vkareh** [comment]
sorry about the delay, I had forgotten about this. I reverted the python version check to >= 2.7 so that it can be built anywhere

**Comment by vkareh** [comment]
@XRevan86, @monsta - If we move forward with this, it doesn't remove the option of porting to GIR in the future, while still getting Python 3 working in the interim. Any chance this can get code reviewed/tested?

**Comment by flexiondotorg** [comment]
I agree, would be good to enable the switch the Python3 and then migrate to GIR in the future.

**Comment by monsta** [comment]
I don't mind this port itself, I just feel it's an additional work that will become unneeded when GI port is done, because these manual bindings won't be needed. Otherwise, I'm fine with it. 🙂 

I can test that Python 2 code still works with the existing mintMenu and Mozo. What do I use for Python 3 test?

**Comment by monsta** [comment]
Ah... I guess it's https://github.com/ubuntu-mate/mate-menu/pull/28 - missed that.

**Comment by monsta** [comment]
> For example, Ubuntu MATE has few applications that use python2 it would be great to get everything transitioned to python3 so the iso size can be better managed. 

Speaking of this, we have eom and pluma with Python 2 plugins, and also python-caja which is needed by several 3rd party extensions shipped with Ubuntu MATE.

**Comment by raveit65** [comment]
> I don't mind this port itself, I just feel it's an additional work that will become unneeded when GI port is done, because these manual bindings won't be needed.

Is some where working on it or is more a wish, or do i miss something ;-)
Personal i don't like to see more build warnings from gir.
I am sorry to say , no one really maintain our gir code in **all** packages.
So, let us merge this.
@XRevan86 
Is this ready to go?

**Comment by monsta** [comment]
It's more like a wish indeed. I didn't see anyone working on GI yet.

File: python/matemenu.c [hunk 1]
```diff
@@ -1945,4 +1991,9 @@ DL_EXPORT(void) initmatemenu(void)
 
 	PyModule_AddIntConstant(mod, "SORT_NAME",         MATEMENU_TREE_SORT_NAME);
 	PyModule_AddIntConstant(mod, "SORT_DISPLAY_NAME", MATEMENU_TREE_SORT_DISPLAY_NAME);
+
+#if PY_MAJOR_VERSION < 3
+  mod = NULL;
+#endif
+  return mod;
```
Comment (monsta) [inline python/matemenu.c]: This `return` should be there only for Python 3, as with Python 2 the function doesn't return anything...

**Comment by monsta** [comment]
Besides the warning about that `return`, there are some more:

```c
matemenu.c:56:0: warning: "PyUnicode_FromString" redefined
  #define PyUnicode_FromString PyString_FromString
 ^
In file included from /usr/include/python2.7/Python.h:85:0,
                 from matemenu.c:22:
/usr/include/python2.7/unicodeobject.h:281:0: note: this is the location of the previous definition
 # define PyUnicode_FromString PyUnicodeUCS4_FromString
 ^
matemenu.c:57:0: warning: "PyUnicode_Check" redefined
  #define PyUnicode_Check PyString_Check
 ^
In file included from /usr/include/python2.7/Python.h:85:0,
                 from matemenu.c:22:
/usr/include/python2.7/unicodeobject.h:427:0: note: this is the location of the previous definition
 #define PyUnicode_Check(op) \
 ^
matemenu.c:1761:0: warning: "PyLong_Check" redefined
  #define PyLong_Check PyInt_Check
 ^
In file included from /usr/include/python2.7/Python.h:88:0,
                 from matemenu.c:22:
/usr/include/python2.7/longobject.h:14:0: note: this is the location of the previous definition
 #define PyLong_Check(op) \
 ^
```
That's when building in Mint 18 (based on Ubuntu 16.04), with Python 2.7 autodetected during the build.

**Comment by monsta** [comment]
In runtime all seems ok, mintMenu works as before.

**Comment by monsta** [comment]
Since we'll drop support for GTK+ 3.18 before 1.20 release, I've decided to build it in Debian Testing as well. It's all the same, Python 2 is the default one. How do you make it build with Python 3 instead?

**Comment by raveit65** [comment]
@vkareh 
We have a open question :-)
https://github.com/mate-desktop/mate-menus/pull/54#issuecomment-333393171

**Comment by vkareh** [comment]
@raveit65 - thanks for reminding me of this! :P
@monsta - I added a few extra checks to avoid redefinition (and so eliminate all the build warnings)

To build with Python 3 you'll need to set the `PYTHON` env var during the configuration phase:
```bash
make clean
PYTHON=python3 ./autogen.sh --prefix=/foo/bar # etc...
make
sudo make install
```

**Comment by raveit65** [comment]
@vkareh 
With using
PYTHON=python3 ./configure --enable-python
i got this configure result.
checking whether python3 version is >= 2.7... yes
checking for python3 version... 3.6
checking for python3 platform... linux
checking for python3 script directory... ${prefix}/lib/python3.6/site-packages
checking for python3 extension module directory... ${exec_prefix}/lib64/python3.6/site-packages
checking for PYTHON... no
checking for gobject-introspection... yes
<cut>
Build python bindings:        no
Build introspection support:  yes

With ./configure --enable-python

checking for python... /usr/bin/python
checking for python version... 2.7
checking for python platform... linux2
checking for python script directory... ${prefix}/lib/python2.7/site-packages
checking for python extension module directory... ${exec_prefix}/lib64/python2.7/site-packages
checking for PYTHON... yes
checking for gobject-introspection... yes
<cut>
Build python bindings:        yes
Build introspection support:  yes

So --enable-python is for python2 only which is a bit confusing.
Can we have a switch for python3 or a automatic switch depends on which python2/3-devel version is installed?

Here in f26 i have both -devel packages installed.
```
[rave@mother ~]$ rpm -qa python3-devel
python3-devel-3.6.2-8.fc26.x86_64
[rave@mother ~]$ rpm -qa python2-devel
python2-devel-2.7.14-2.fc26.x86_64
```
But using only ./configure gives me that result
checking for a Python interpreter with version >= 2.7... python
checking for python... /usr/bin/python
checking for python version... 2.7
checking for python platform... linux2
checking for python script directory... ${prefix}/lib/python2.7/site-packages
checking for python extension module directory... ${exec_prefix}/lib64/python2.7/site-packages
checking for PYTHON... yes
<cut>
Build python bindings:        yes
Build introspection support:  yes

Well, i need to investigate what i should do in SPEC file for rpm.

**Comment by raveit65** [comment]
With using 'export PYTHON=python3'
I was able to build a rpm with python3. Only adding 
BuildRequires:  python3-devel
wasn't sufficient. Not really comfortable.

**Comment by raveit65** [comment]
I was able to rebuild mate-menu and the port to python3 seems to be fine, but other apps like mozo don't work anymore, looks like mozo needs ported to python3 too.
```
[rave@mother ~]$ mozo
Traceback (most recent call last):
  File "/usr/bin/mozo", line 22, in <module>
    from Mozo.MainWindow import MainWindow
  File "/usr/lib/python2.7/site-packages/Mozo/MainWindow.py", line 19, in <module>
    import matemenu
ImportError: No module named matemenu
```
So we need to upgrade all apps in one rush, maybe for MATE-1.20 release in a few months.
@monsta @XRevan86 ^^^^^^?
Also, other apps like mintmenu from other project are involved.
For fedora i will drop mintmenu soon if mate-menu will be developed further.
@flexiondotorg ^^^^ ?

**Comment by vkareh** [comment]
I can work on mozo. Do we need to work on MenuLibre as well @flexiondotorg? As far as I know mozo is the default menu editor, but MenuLibre seems to be a supported alternative in Ubuntu.

**Comment by raveit65** [comment]
> I can work on mozo. 

+1
> Do we need to work on MenuLibre as well

Not sure, i don't know it very well, but isn't it use by differents DEs, ie. xfce and others?

**Comment by lukefromdc** [comment]
When I first started using MATE, I had one reason to use mintmenu: the fact that it could be themed independently of the GTK theme. Only thing was, the pycompile job at session start bogged down the netbook that was one of the machines it was used on. 

Well. we've had that theme support in mate-panel since version 1.12, so that part would only need some kind of gui feature that would write to ~/config/gtk-3.0/gtk.css (using .mate-panel-menu-bar to limit it to panel menus) to do the exact same thing. I have not been able to figure out how to do that though.

The search feature is the other major thing mintmenu (and cairo-dock's menu) have that we don't have natively. Not one I've ever used but might be one of the reasons many use Mintmenu.

**Comment by raveit65** [comment]
Mate-menu is a fork of mintmenu.
https://github.com/ubuntu-mate/mate-menu/pull/28
And it does exactly the same.

**Comment by monsta** [comment]
If we port these Python apps (mintMenu, mate-menu, mozo), they will be made Python 3 only. But then we wouldn't need any switches when building this library... if there won't be any Python 2 apps left that depend on python-mate-menu.

**Comment by ghostplant** [comment]
I want this feature as well and currently I cannot install mate-menus without installing python2.7-* on Ubuntu 18.04.

**Comment by monsta** [comment]
This isn't the only MATE part which currently needs Python 2. We have Python plugins support in eom and Pluma, and python-caja for Python extensions support in Caja. Ubuntu has some of these extensions: deja-dup, caja-rename, folder-color...

**Comment by monsta** [comment]
> I can work on mozo. Do we need to work on MenuLibre as well @flexiondotorg? As far as I know mozo is the default menu editor, but MenuLibre seems to be a supported alternative in Ubuntu.

@vkareh: yes, it's supported, but it doesn't use this library. I see it depends on gnome-menus, and it also uses Python 3. So nothing to do there 🙂

**Comment by monsta** [comment]
I found out that Cinnamon's own menu editor is forked from Alacarte, just like Mozo (https://github.com/linuxmint/Cinnamon/commit/afd1225ec584ac5c028ef303ed3be8f83c794428), and it's now ported to Python 3 (https://github.com/linuxmint/Cinnamon/pull/7074). Might help us in porting Mozo.

**Comment by raveit65** [comment]
Closing this as other PR is merged.

**Pull request closed without merging**